File: typescript/private_key/7_market_sell_order.ts
/**
 * 7_market_sell_order.ts — FOK market SELL of full YES token balance (Iran market).
 *
 * EOA-based: reads YES token balance from CTF contract, sells full amount at best bid.
 * Run 6_market_buy_order.ts first to have YES tokens.
 *
 * Run:
 *   npm run 7:msell
 */

import "dotenv/config";
import { createWalletClient, createPublicClient, http, erc20Abi, formatUnits } from "viem";
import { polygon } from "viem/chains";
import { privateKeyToAccount } from "viem/accounts";
import { ClobClient, Side, OrderType, type UserMarketOrderV2 } from "@polymarket/clob-client-v2";

const PRIVATE_KEY = process.env.PRIVATE_KEY!;
const CLOB_HOST   = process.env.CLOB_V2_BASE_URL ?? "https://clob-v2.polymarket.com";
const RPC_URL     = process.env.POLYGON_RPC_URL  ?? "https://polygon-bor-rpc.publicnode.com";
const EVENT_ID    = "73106";
const PUSD        = "0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB" as `0x${string}`;
const CTF         = "0x4D97DCd97eC945f40cF65F87097ACe5EA0476045" as `0x${string}`;

const CTF_BALANCE_ABI = [{
  name: "balanceOf", type: "function", stateMutability: "view",
  inputs: [{ name: "account", type: "address" }, { name: "id", type: "uint256" }],
  outputs: [{ name: "", type: "uint256" }],
}] as const;

const account      = privateKeyToAccount(`0x${PRIVATE_KEY}` as `0x${string}`);
const walletClient = createWalletClient({ account, chain: polygon, transport: http(RPC_URL) });
const publicClient = createPublicClient({ chain: polygon, transport: http(RPC_URL) });
const EOA          = account.address;

console.log("EOA  :", EOA);
console.log("CLOB :", CLOB_HOST, "\n");

// ── Init ClobClient V2 (EOA) ──────────────────────────────────────────────────

const tempClient = new ClobClient({ host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0 });
const creds      = await (tempClient as any).createOrDeriveApiKey();
const client     = new ClobClient({
  host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0,
  creds: { key: creds.key, secret: creds.secret, passphrase: creds.passphrase },
});
console.log("API key :", creds.key, "✓");

// ── Market info ───────────────────────────────────────────────────────────────

const event    = await fetch(`https://gamma-api.polymarket.com/events/${EVENT_ID}`).then(r => r.json()) as any;
const condId   = event.markets?.[0]?.conditionId as string;
const info     = await (client as any).getClobMarketInfo(condId);
const tokens   = (info?.t ?? []) as { t: string; o: string }[];
const tokenId  = tokens.find(t => t.o === "Yes")!.t;
const tickSize = (info?.mts?.toString() ?? "0.01") as "0.1" | "0.01" | "0.001" | "0.0001";
const negRisk  = info?.nr ?? false;

console.log(`Market: tick=${tickSize}  negRisk=${negRisk}  token=${tokenId.slice(0, 16)}…`);

// ── YES token balance (ERC-1155) ──────────────────────────────────────────────

const rawYes    = await publicClient.readContract({ address: CTF, abi: CTF_BALANCE_ABI, functionName: "balanceOf", args: [EOA, BigInt(tokenId)] });
const yesShares = Number(rawYes) / 1e6;
const rawPusd   = await publicClient.readContract({ address: PUSD, abi: erc20Abi, functionName: "balanceOf", args: [EOA] });
const pusdBefore = parseFloat(formatUnits(rawPusd, 6));

console.log(`YES balance : ${yesShares.toFixed(6)} shares`);
console.log(`pUSD before : $${pusdBefore.toFixed(6)}`);

if (yesShares < 0.01) { console.error("⚠ No YES tokens — run 6_market_buy_order.ts first."); process.exit(1); }

// ── Book snapshot ─────────────────────────────────────────────────────────────

const book  = await (client as any).getOrderBook(tokenId);
const bids  = [...(book?.bids ?? [])].sort((a: any, b: any) => parseFloat(b.price) - parseFloat(a.price));
if (bids.length === 0) { console.error("⚠ No bids — can't market sell."); process.exit(1); }
const bestBid = parseFloat(bids[0].price);
console.log(`Best bid : $${bestBid}  Selling: ${yesShares.toFixed(6)} shares`);

// ── FOK market SELL (full YES balance) ───────────────────────────────────────

console.log(`\nPlacing FOK SELL: ${yesShares.toFixed(6)} shares @ $${bestBid}…`);
const sellOrder: UserMarketOrderV2 = { tokenID: tokenId, price: bestBid, amount: yesShares, side: Side.SELL };
const sellResp = await client.createAndPostMarketOrder(sellOrder, { tickSize, negRisk }, OrderType.FOK);
console.log(`  Order ID : ${sellResp.orderID}  status: ${sellResp.status}`);
if (!sellResp.orderID) { console.error("Sell rejected:", (sellResp as any).errorMsg); process.exit(1); }

// ── Poll fill ─────────────────────────────────────────────────────────────────

console.log("\nPolling for fill (up to 30 s)…");
let filled = false;
for (let i = 0; i < 30; i++) {
  await new Promise(r => setTimeout(r, 1000));
  const open: any[] = await (client as any).getOpenOrders() ?? [];
  if (!open.find((o: any) => o.id === sellResp.orderID)) { filled = true; break; }
  process.stdout.write(".");
}
console.log(filled ? "\n✓ Filled" : "\n⚠ Still open after 30 s");

// ── Poll balance until settlement (10 × 3 s) ─────────────────────────────────

const expectedPusd = pusdBefore + yesShares * bestBid;
console.log(`\nExpected pUSD (optimistic): $${expectedPusd.toFixed(6)}`);
console.log("Polling for settlement (10 × 3 s)…");

let yesAfter = yesShares, pusdAfter = pusdBefore;
for (let i = 0; i < 10; i++) {
  await new Promise(r => setTimeout(r, 3000));
  const [rawYesAfter, rawPusdAfter] = await Promise.all([
    publicClient.readContract({ address: CTF, abi: CTF_BALANCE_ABI, functionName: "balanceOf", args: [EOA, BigInt(tokenId)] }),
    publicClient.readContract({ address: PUSD, abi: erc20Abi, functionName: "balanceOf", args: [EOA] }),
  ]);
  yesAfter  = Number(rawYesAfter) / 1e6;
  pusdAfter = parseFloat(formatUnits(rawPusdAfter, 6));
  console.log(`  [${i + 1}/10] YES: ${yesAfter.toFixed(4)}  pUSD: $${pusdAfter.toFixed(4)}`);
  if (pusdAfter >= expectedPusd * 0.95) { console.log("✓ Settlement confirmed"); break; }
}

console.log(`\nYES after   : ${yesAfter.toFixed(6)} shares`);
console.log(`pUSD before : $${pusdBefore.toFixed(6)}`);
console.log(`pUSD after  : $${pusdAfter.toFixed(6)}`);
console.log(`Net PnL     : $${(pusdAfter - pusdBefore).toFixed(6)}`);

File: python/private_key/7_market_sell_order.py
"""
7_market_sell_order.py — FOK market SELL of full YES token balance.

EOA-based: reads YES token balance from CTF (ERC-1155), sells full amount at best bid.
Run 6_market_buy_order.py first to have YES tokens.

Run:
    python 7_market_sell_order.py

Requires: PRIVATE_KEY, CLOB_V2_BASE_URL, POLYGON_RPC_URL in .env
"""

import os
import time
import requests
from dotenv import load_dotenv
from web3 import Web3
from py_clob_client_v2.client import ClobClient
from py_clob_client_v2.clob_types import MarketOrderArgsV2, PartialCreateOrderOptions, OrderType, BalanceAllowanceParams, AssetType

load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), ".env"))

PRIVATE_KEY = os.environ["PRIVATE_KEY"]
CLOB_HOST   = os.environ.get("CLOB_V2_BASE_URL", "https://clob-v2.polymarket.com")
RPC_URL     = os.environ.get("POLYGON_RPC_URL", "https://polygon-bor-rpc.publicnode.com")
EVENT_ID    = "73106"
PUSD        = Web3.to_checksum_address("0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB")
CTF         = Web3.to_checksum_address("0x4D97DCd97eC945f40cF65F87097ACe5EA0476045")

ERC20_ABI = [{"name": "balanceOf", "type": "function", "stateMutability": "view",
              "inputs": [{"name": "account", "type": "address"}],
              "outputs": [{"name": "", "type": "uint256"}]}]

CTF_ABI = [{"name": "balanceOf", "type": "function", "stateMutability": "view",
            "inputs": [{"name": "account", "type": "address"}, {"name": "id", "type": "uint256"}],
            "outputs": [{"name": "", "type": "uint256"}]}]

w3  = Web3(Web3.HTTPProvider(RPC_URL))
eoa = w3.eth.account.from_key("0x" + PRIVATE_KEY).address
pusd_contract = w3.eth.contract(address=PUSD, abi=ERC20_ABI)
ctf_contract  = w3.eth.contract(address=CTF,  abi=CTF_ABI)

# ── Init client ───────────────────────────────────────────────────────────────
temp_client = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, signature_type=0)
creds       = temp_client.create_or_derive_api_key()
client      = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, creds=creds, signature_type=0)
print(f"EOA     : {eoa}")
print(f"API key : {creds.api_key} ✓\n")

# ── Market info ───────────────────────────────────────────────────────────────
event    = requests.get(f"https://gamma-api.polymarket.com/events/{EVENT_ID}").json()
cond_id  = event["markets"][0]["conditionId"]
info     = client.get_clob_market_info(cond_id)
tokens   = info.get("t", [])
token_id = next((t["t"] for t in tokens if t["o"] == "Yes"), "")
tick_size = str(info.get("mts", "0.01"))
neg_risk  = info.get("nr", False)
options  = PartialCreateOrderOptions(tick_size=tick_size, neg_risk=neg_risk)

print(f"Market: tick={tick_size}  negRisk={neg_risk}  token={token_id[:16]}…")

# ── Sync YES token balance/allowance to CLOB server ──────────────────────────
client.update_balance_allowance(BalanceAllowanceParams(asset_type=AssetType.CONDITIONAL, token_id=token_id))
print("Conditional balance synced ✓\n")

# ── YES token balance (ERC-1155) ──────────────────────────────────────────────
raw_yes    = ctf_contract.functions.balanceOf(eoa, int(token_id)).call()
yes_shares = int(raw_yes / 1e6 * 100) / 100  # floor to 2 decimal places
raw_pusd   = pusd_contract.functions.balanceOf(eoa).call()
pusd_before = raw_pusd / 1e6

print(f"YES balance  : {yes_shares:.6f} shares")
print(f"pUSD before  : ${pusd_before:.6f}")

if yes_shares < 0.01:
    print("⚠ No YES tokens — run 6_market_buy_order.py first."); exit(1)

# ── Order book — best bid ─────────────────────────────────────────────────────
book = client.get_order_book(token_id)
bids = sorted(book["bids"], key=lambda b: float(b["price"]), reverse=True)
if not bids:
    print("⚠ No bids — can't market sell."); exit(1)

best_bid = float(bids[0]["price"])
print(f"Best bid : ${best_bid}  Selling: {yes_shares:.6f} shares\n")

# ── FOK market SELL (amount = shares to sell) ────────────────────────────────
print(f"Placing FOK SELL: {yes_shares:.6f} shares @ ${best_bid}…")
sell_resp = client.create_and_post_market_order(
    MarketOrderArgsV2(token_id=token_id, amount=yes_shares, side="SELL", price=best_bid),
    options, OrderType.FOK,
)
sell_id = sell_resp.get("orderID") if isinstance(sell_resp, dict) else sell_resp.orderID
status  = sell_resp.get("status")  if isinstance(sell_resp, dict) else sell_resp.status
print(f"  Order ID : {sell_id}  status: {status}")
if not sell_id:
    print(f"  Rejected : {sell_resp.get('errorMsg') if isinstance(sell_resp, dict) else sell_resp}"); exit(1)

# ── Poll fill ─────────────────────────────────────────────────────────────────
print("\nPolling for fill (up to 30 s)…")
filled = False
for i in range(30):
    time.sleep(1)
    open_orders = client.get_open_orders() or []
    ids = [o.get("id", o.get("orderID", "")) for o in open_orders]
    if sell_id not in ids:
        filled = True
        break
    print(".", end="", flush=True)
print(f"\n{'✓ Filled' if filled else '⚠ Still open after 30 s'}")

# ── Poll balance until settlement (10 × 3 s) ─────────────────────────────────
expected_pusd = pusd_before + yes_shares * best_bid
print(f"\nExpected pUSD (optimistic): ${expected_pusd:.6f}")
print("Polling for settlement (10 × 3 s)…")

yes_after = yes_shares
pusd_after = pusd_before
for i in range(10):
    time.sleep(3)
    raw_yes_after  = ctf_contract.functions.balanceOf(eoa, int(token_id)).call()
    raw_pusd_after = pusd_contract.functions.balanceOf(eoa).call()
    yes_after  = raw_yes_after  / 1e6
    pusd_after = raw_pusd_after / 1e6
    print(f"  [{i + 1}/10] YES: {yes_after:.4f}  pUSD: ${pusd_after:.4f}")
    if pusd_after >= expected_pusd * 0.95:
        print("✓ Settlement confirmed")
        break

print(f"\nYES after   : {yes_after:.6f} shares")
print(f"pUSD before : ${pusd_before:.6f}")
print(f"pUSD after  : ${pusd_after:.6f}")
print(f"Net PnL     : ${pusd_after - pusd_before:.6f}")
